gazebo_ros/src/executor.cpp

```diff
@@ -22,15 +22,22 @@ namespace gazebo_ros
 Executor::Executor()
 : spin_thread_(std::bind(&Executor::run, this))
 {
+  using namespace std::chrono_literals;
   sigint_handle_ = gazebo::event::Events::ConnectSigInt(std::bind(&Executor::shutdown, this));
+  while (!this->spinning) {
+    // TODO(ivanpauno): WARN Terrible hack here!!!!
+    // We cannot call rclcpp::shutdown asynchronously, because it generates exceptions that
+    // cannot be caught properly
+    // Executor::cancel() doesn't cause this problem, but it has a race.
+    // Wait until the launched thread starts spinning to avoid the race ...
```

Comment:
Cool, I didn't know about `cancel()`. I guess the race is if the executor is constructed and quickly destructed. Instead of sleeping here, alternatively, we could block in the destructor using an additional flag (we are already blocking there with `spin_thread_.join()`), e.g.

```c++
Executor::~Executor()
{
  while (!this->thread_done_) {
    this->shutdown();
    // Or maybe std::condition_variable would be slightly better?
    std::this_thread::sleep_for(100ms);
  }
  spin_thread_.join();
}

void Executor::run()
{
  spin();
  this->thread_done_ = true;
}
```

I don't think we have to worry about asynchronous calls to `Executor::shutdown()` since it is private.

Either way, it's still a not a great hack.

Reply:
>     // Or maybe std::condition_variable would be slightly better?

If you use a condition variable, you still have a race because of the time that happens between you trigger the condition variable and you call spin.

>  Instead of sleeping here, alternatively, we could block in the destructor using an additional flag (we are already blocking there with spin_thread_.join()), e.g.

Sounds good to me

Reply:
>> Instead of sleeping here, alternatively, we could block in the destructor using an additional flag (we are already blocking there with spin_thread_.join()), e.g.
>
> Sounds good to me

oh wait, the private `shutdown` method can be called from a signal handler, and calling sleep there doesn't sound valid (and we want the signal handler to cancel the executor).

So, I guess that I will keep blocking in the constructor (which sounds like safe to do).

Reply:
> If you use a condition variable, you still have a race because of the time that happens between you trigger the condition variable and you call spin.

I imagined the trigger happens after `spin()`, e.g.

```
void Executor::run()
{
  spin();
  this->cv_.notify_all();
}
```

And then we condition the while-loop in the destructor on the condition variable.

Reply:
> I imagined the trigger happens after spin(), e.g.

`spin` is blocking, so that wouldn't block.
You could try to create a "one time timer" and trigger the condition variable there, but really that's a lot of boilerplate and I'm not sure if it has some caveats or not.

Reply:
:thinking: I was imagining the trigger happens at the end of the spin thread. It seems like it would work to me, or maybe I'm misunderstanding something about condition variables.

`spin()` blocks until the executor is canceled.
Then the spin thread notifies the main thread via the condition variable.

Meanwhile, the main thread is blocking in the destructor until the spin thread is done.

Reply:
err, nevermind, I got it backwards. I see the issue :upside_down_face: